add generateFilterOptions function

Alexandru-Bulai opened the pull request:
# PR 31-01-24
## Overview
In this PR we adding a function for the user input on the filters options
## Key Changes :
 1) gallery.test.js :
    - Added generateFilterOptions function that allow user to select only one valid option
    - If user didnt select a valid option then we return a null value and it will not be applied

2) gallery,js :
    - Moved generateFilterOptions to gallery.js file after finishing testing and exported the funciton to gallery.test.js
# Additional notes
I had to slightly modify the test for invalid rating input and invalid gender input because they were actuallly valid inputs.

<img width="1501" alt="Screenshot 2024-01-31 at 10 35 46" src="https://github.com/Alexandru-Bulai/website-project/assets/114404620/63954f69-e4b1-44b2-be71-a56c2f5b96e5">

Alexandru-Bulai committed:
add generateFilterOptions function

vladiiancu commented on test/gallery.test.js:
```diff
@@ -109,7 +95,7 @@ describe('Applying Gallery Filters - allow only 1 value per filter', () => {
   test('if user selects invalid rating, it will not apply it', () => {
     let userSelectedFilters = { rating: [-5] }
     expect(generateFilterOptions(userSelectedFilters)).toHaveProperty('rating', null)
-    userSelectedFilters = { rating: 3 }
```
Well this one is actually good. 
The point is that the input is invalid because it is not an array. 

It was specifically put as a valid number so that the test is even more relevant.

Alexandru-Bulai replied:
Right, i complitely missed that... will fix.

vladiiancu commented on test/gallery.test.js:
```diff
@@ -87,7 +73,7 @@ describe('Applying Gallery Filters - allow only 1 value per filter', () => {
   test('if user selected gender filters contains invalid value, it will not apply that filter', () => {
     let userSelectedFilters = { type: ['cat'], gender: ['extraterestru'] }
     expect(generateFilterOptions(userSelectedFilters)).toHaveProperty('gender', null)
-    userSelectedFilters = { type: ['cat'], gender: 'male' }
```
Again. The input is invalid because it is not an arrest. Doesn't matter if the actual value is correct or not.
So this change is not required as well

Alexandru-Bulai committed:
lintern fix

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -62,4 +62,38 @@ function reportGalleryStats (imagesInDom) {
+        const lastFilterValue = filterValue[filterValue.length - 1]
+        // Check if the last value of the array is an allowed value
+        activeFilters[filterKey] = allowedFilters[filterKey].includes(lastFilterValue) ? lastFilterValue : null
+      } else {
```
Transform this into a guard block.

Start the function by checking all conditions. Start with the known failing cases. Return early. 

If it passes each guard then at the end of the function it reaches it means that all conditions are met do something you are supposed to do

Alexandru-Bulai committed:
guard block, early returns

Alexandru-Bulai committed:
guard block comment

Alexandru-Bulai committed:
lintern comments fix

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -62,4 +62,44 @@ function reportGalleryStats (imagesInDom) {
+/**
+ *
+ * @param {Array} userSelectedFilters
+ * @returns {Object} activeFilters
```
This says that the function returns on object.
With the recent changes, the functions does more than this.

In some cases it returns.
In others it sets activefilters object to something and then returns null.

And in others return an object.

vladiiancu commented on public/JavaScript/gallery.js:
```diff
@@ -62,4 +62,44 @@ function reportGalleryStats (imagesInDom) {
+    // If filter option is not in the allowed array set it too null
+    const filterValues = userSelectedFilters[filterKey]
+    if (!Array.isArray(filterValues)) {
+      activeFilters[filterKey] = null
```
Why? 
It is initialised with null.
Is it changed during the execution of this function?

Alexandru-Bulai replied:
Right they were already declared null, that line is useless there

Alexandru-Bulai committed:
return object if valid input else return null

Alexandru-Bulai committed:
lintern fix

vladiiancu committed:
refactor generateFilterOptions
We had a bunch of returns of the same variable on each branch. Refactoring this to include a single return only.

vladiiancu committed:
refactor generateFilterOptions again
This time we switched to using Sets as well as optional chaining inside the if
